solutions/CF Garland.mdx

```diff
@@ -2,13 +2,141 @@
 id: cf-garland
 source: CF
 title: An impassioned circulation of affection
-author: Neo Wang
+author: Neo Wang, Qi Wang, Jesse Choe
 ---
 
 [Official Editorial](https://codeforces.com/blog/entry/52449)
 
+## Solution (Two Pointers)
+
+**Time Complexity:** $\mathcal{O}(nq)$
+
+For each query we can keep a sliding window where no more than $a$ letters in the window are not $b$.
+
+<LanguageSection>
+
+<CPPSection>
+
+### C++ Implementation:
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+using ll = long long;
+using vi = vector<int>;
+#define pb push_back
+#define all(x) begin(x), end(x)
+#define sz(x) (int)(x).size()
+using pi = pair<int,int>;
+#define f first
+#define s second
+#define mp make_pair
+void setIO(string name = "") {
+	cin.tie(0)->sync_with_stdio(0); // see /general/fast-io
+	if (sz(name)) {
+		freopen((name+".in").c_str(), "r", stdin); // see /general/io
+		freopen((name+".out").c_str(), "w", stdout);
+	}
+}
+
+int N, Q;
+string W;
+
+int main() {
+	setIO();
+	cin >> N >> W >> Q;
+
+	for(int i = 0; i < Q; i++){
+		int a, m = 0; char b;
+		cin >> a >> b;
+		
+		int l, r;
+		l = r = 0;
+		while(l < N && r < N) {
+			while(r < N) {
+				if(W[r] != b) {
+					if(a == 0) break;
+					a--;
+				}
+				r++;
+			}
+
+			m = max(m, r - l);
+
+			a += W[l++] != b;
+		}
+
+		cout << m << endl;
+	}
+}
+```
+
+</CPPSection>
+
+</LanguageSection>
+
+## Alternative Solution (Two Pointers with Prefix Sums)
+
+**Time Complexity:** $\mathcal{O}(nq)$
+
+For each query $q$ we can use prefix sums to check the number of occurrences of a given letter $c_i$ and use two pointers to find the maximum Koyomity ending at a given index $j$ from $0 \dots n$.
+
+<LanguageSection>
+
+<CPPSection>
+
+### C++ Implementation:
+
+```cpp
+#include <bits/stdc++.h>
+
+using namespace std;
+
+int dp[26][1501];
+
+int main(){
+	int n; cin >> n;
+	string s; cin >> s;
+	for(int i=1; i<=n; i++){
+		int color = s[i-1] - 'a';
+		dp[color][i]++;
+	}
+	for(int i=0; i<26; i++){
+		for(int j=1; j<=n; j++){
+			dp[i][j] = dp[i][j-1] + dp[i][j];
+		}
+	}
+	int q; cin >> q;
+	for(int i=0; i<q; i++){
+		int m; char c; cin >> m >> c;
+		c -= 'a';
+		int p1 = 1, p2 = 1, koyomity = 1;
+		while(p2 <= n){
+			int c_count = dp[c][p2] - dp[c][p1-1] + m;
+			if((c_count) < (p2 - p1 + 1)){
+				p1++;
+			} else {
+				koyomity = max(koyomity, (p2 - p1 + 1));
+				p2++;
+			}
+		}
+		cout << koyomity << endl;
+	}
+}
+```
+
+</CPPSection>
+
+</LanguageSection>
+
 ## Alternative Solution (Dynamic Programming)
 
+<Warning>
+
+Dynamic programming is a [Gold](/gold/intro-dp) topic.
+
+</Warning>
+
 **Time Complexity:** $\mathcal{O}(n^2k + q)$
 
 We are performing up to $200\,000$ queries, so we must have some way to quickly precompute our answer in constant time.
```

Comment:
```suggestion
We are performing up to $200\,000$ queries, so it may be optimal to have some way to quickly precompute our answer in constant time.
```